File: JavaProject/GrowNet/src/main/java/ai/nektron/grownet/Neuron.java
package ai.nektron.grownet;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.function.BiConsumer;

import static java.lang.Math.abs;

/**
 * Base neuron with slot logic and unified onInput/onOutput contract.
 * Subclasses override fire(...) for excitatory / inhibitory / modulatory behaviour.
 */
public class Neuron {

    // -------- state --------
    protected final String neuronId;
    protected final LateralBus bus;

    // slot memory: key = slot id, val = weight
    protected final Map<Integer, Weight> slots = new HashMap<>();

    // explicit synapses (kept for future, even if Tract handles layer–layer routing)
    protected final List<Synapse> outgoing = new ArrayList<>();

    // slot policy + optional cap
    protected final SlotEngine slotEngine;
    protected final int       slotLimit;   // -1 = unbounded

    // last input (for %delta binning)
    protected boolean haveLastInput = false;
    protected double  lastInputValue = 0.0;

    // Temporal focus (anchor-based)
    protected double  focusAnchor = 0.0;
    protected boolean focusSet = false;
    protected long    focusLockUntilTick = 0L;

    // NEW: hooks notified whenever this neuron fires
    private final List<BiConsumer<Double, Neuron>> fireHooks = new ArrayList<>();

    // -------- construction --------
    public Neuron(String neuronId, LateralBus bus, SlotConfig slotConfig, int slotLimit) {
        this.neuronId   = neuronId;
        this.bus        = bus;
        this.slotEngine = new SlotEngine(slotConfig);
        this.slotLimit  = slotLimit;
    }

    // -------- wiring (optional, still useful for unit tests) --------
    public Synapse connect(Neuron target, boolean feedback) {
        Synapse s = new Synapse(target, feedback);
        outgoing.add(s);
        return s;
    }

    // -------- IO --------
    /**
     * Route a scalar in, learn locally, maybe fire. Returns true if fired.
     */
    public boolean onInput(double value) {
        // choose / create slot
        final int slotId = haveLastInput
                ? slotEngine.slotId(lastInputValue, value, slots.size())
                : 0; // imprint
        Weight slot = slots.get(slotId);
        if (slot == null) {
            if (slotLimit >= 0 && slots.size() >= slotLimit) {
                // simple reuse policy: recycle the first slot
                int first = slots.keySet().iterator().next();
                slot = slots.get(first);
            } else {
                slot = new Weight();
                slots.put(slotId, slot);
            }
        }

        // learn under neuromodulation
        slot.reinforce(bus.getModulationFactor());

        // update the adaptive threshold and decide
        boolean fired = slot.updateThreshold(value);
        if (fired) {
            fire(value);                // subtype behaviour
            notifyFireHooks(value);     // <— Tract listens here
        }

        haveLastInput  = true;
        lastInputValue = value;
        return fired;
    }

    /**
     * Default no‑op; OutputNeuron overrides to expose values externally.
     */
    public void onOutput(double amplitude) { /* no‑op by default */ }

    /**
     * Subclasses override to implement excitatory / inhibitory / modulatory semantics.
     * Base keeps the unified signature but does not enforce behaviour.
     */
    protected void fire(double inputValue) { /* handled by subclasses */ }

    // -------- small introspection used by demos --------
    public double neuronValue(String mode) {
        if (slots.isEmpty()) return 0.0;

        switch (mode) {
            case "readiness": {
                double best = -1e300;
                for (Weight weight : slots.values()) {
                    double margin = weight.getStrengthValue() - weight.getThresholdValue();
                    if (margin > best) best = margin;
                }
                return best;
            }
            case "firing_rate": {
                double sum = 0.0;
                for (Weight weight : slots.values()) {
                    sum += weight.getEmaRate();
                }
                return sum / slots.size();
            }
            case "memory": {
                double acc = 0.0;
                for (Weight weight : slots.values()) {
                    acc += Math.abs(weight.getStrengthValue());
                }
                return acc;
            }
            default:
                return neuronValue("readiness");
        }
    }

    // Neuron.java  — add inside class Neuron
    /**
     * Remove outgoing synapses that are both stale and weak.
     *
     * @param staleWindow        prune if (currentStep - lastStep) > staleWindow
     * @param minStrength        prune if weight.strength < minStrength
     * @return number of synapses removed
     */
    public int pruneSynapses(long staleWindow, double minStrength) {
        int removed = 0;
        final long currentStep = bus.getCurrentStep();   // LateralBus (layer bus) step

        // Keep list to avoid ConcurrentModificationException
        final List<Synapse> keep = new ArrayList<>(outgoing.size());
        for (Synapse synapse : outgoing) {
            final long    lastStep      = synapse.getLastStep();
            final double  strength      = synapse.getWeight().getStrengthValue();
            final boolean isStale       = (currentStep - lastStep) > staleWindow;
            final boolean isWeak        = strength < minStrength;

            if (isStale && isWeak) {
                removed += 1;
                // drop this synapse
            } else {
                keep.add(synapse);
            }
        }
        outgoing.clear();
        outgoing.addAll(keep);
        return removed;
    }

    
    // -------- hooks API (used by Tract) --------

    /** Register a hook (value, who) called every time this neuron fires. */
    public void registerFireHook(BiConsumer<Double, Neuron> hook) {
        if (hook != null) fireHooks.add(hook);
    }

    /** Bridge if you kept a custom functional interface FireHook. */
    public void registerFireHook(FireHook hook) {
        if (hook != null) fireHooks.add((v, who) -> hook.onFire(v, who));
    }

    private void notifyFireHooks(double amplitude) {
        for (BiConsumer<Double, Neuron> h : fireHooks) {
            try { h.accept(amplitude, this); } catch (Exception ignored) { }
        }
    }

    // -------- getters --------
    public Map<Integer, Weight> getSlots()   { return slots; }
    public List<Synapse>        getOutgoing(){ return outgoing; }
    public LateralBus           getBus()     { return bus; }
    public String getId()         { return neuronId; }

    public double getLastInputValue() {
        return lastInputValue;
    }

    public double getFocusAnchor() { return focusAnchor; }
}

File: grownet_python_mojo/python/neuron.py
from typing import List, Callable, Optional, TYPE_CHECKING
from region_bus import RegionBus
from synapse import Synapse

class Neuron:
    """
    Base neuron with a thin event‑style API.
    Subclasses override on_input(...) for specific behavior.
    """
    def __init__(self, neuron_id: str, bus: RegionBus, slot_limit: int = 64) -> None:
        self._id = str(neuron_id)
        self._bus = bus
        self._slot_limit = int(slot_limit)
        self._outgoing: List[Synapse] = []
        self._slots: List[float] = []  # placeholder for "slot engine" accounting
        self._fire_hook: Optional[Callable[["Neuron", float], None]] = None
        self._last_input: float = 0.0

    # -------- connections --------
    def connect(self, other: "Neuron", feedback: bool = False) -> None:
        self._outgoing.append(Synapse(self, other, weight=1.0, feedback=feedback))

    # -------- runtime --------
    def on_input(self, value: float) -> bool:
        """
        Default behavior: store input (optionally gated by inhibition) and fire.
        Returns True if fired (for hooks/tests).
        """
        inhib = self._bus.get_inhibition_factor()
        effective = value * (1.0 - max(0.0, min(1.0, inhib)))
        self._last_input = effective
        return self.fire(effective)

    def fire(self, value: float) -> bool:
        if self._fire_hook:
            self._fire_hook(self, value)
        for s in self._outgoing:
            s.transmit(value)
        return True

    def end_tick(self) -> None:
        self._last_input = 0.0

    # -------- hooks / accessors --------
    def register_fire_hook(self, hook: Callable[["Neuron", float], None]) -> None:
        self._fire_hook = hook

    def get_id(self) -> str:
        return self._id

    def get_outgoing(self) -> List[Synapse]:
        return self._outgoing

    def get_slots(self) -> List[float]:
        return self._slots

    def get_bus(self) -> RegionBus:
        return self._bus


class ExcitatoryNeuron(Neuron):
    pass


class InhibitoryNeuron(Neuron):
    def on_input(self, value: float) -> bool:
        # Flip sign to approximate inhibitory behavior
        return super().on_input(-abs(value))


class ModulatoryNeuron(Neuron):
    def on_input(self, value: float) -> bool:
        # Use modulation to scale the outgoing signal
        m = self._bus.get_modulation_factor()
        return super().on_input(value * (1.0 + m))
